### DISCUSSION TITLE
Bidirectional mavlink communication without using QGroundStation and ArduCopter

### DISCUSSION THREAD

**Discussion by SzateX** [svpcom/wfb-ng#116] [Questions]
Hi, I'm trying to communicate my drone with ground station. Drone should send video data, and mavlink messages to GS, and GS should send commands into drone. I'm not using QGroundStation and ArduCopter, I'm using my own software. 

Now working communication from Drone to GS.
The drone settings are set to:
```
[drone_mavlink]
peer = 'listen://0.0.0.0:14550'

[drone_video]
peer = 'listen://0.0.0.0:5602
```

and GS settings are set to:
```
[gs_mavlink]
peer = 'connect://127.0.0.1:14550'

[gs_video]
peer = 'connect://127.0.0.1:5602'
````

On Drone I'm trying to bind 14550 UDP Port for reading mavlink communicates, but i'm unable to do it.
Have you some advices for make my communication two way?

**Comment by lucasgrelaud** [comment]
Could you test the connection with the `wfb-cli` tool to check if the link is established ?
You should a least see activity on the IP tunnel if the connection between the drone and gs is properly established.
I had issue in the past with mismatched drone and gs keys and this tool help me to figure it out.

Also, could you send us the log generated by the systemd service daemon?
You can access it using "sudo journalctl -u wifibroadcast@gs.service"

**Reply by SzateX** [reply]
In log rather not special: [](https://gist.github.com/SzateX/9e825d4ed2562fcc09f54879f145ea9a).

wfb-cli shows me that connection works ok. The video shows as well without any problems, and mavlink communication from drone to gs works as well... Problem is when I'm trying to bind 15550 port on UDP to read messages on Drone I got: Port already in use. And I have no idea how solve this problem. I need two way communication via mavlink, because i have to set my camera settings by mavlink

**Comment by svpcom** [comment]
```
[drone_mavlink]
peer = 'listen://0.0.0.0:14550'
```
`listen` means that WFB is listening (bind to) UDP port. Add it as endpoint to mavlink router (`mavlink-routerd -e 127.0.0.1:14550 /dev/ttySxx:uart_speed`).

 If you want WFB to connect to some port, use `connect://` url

**Reply by SzateX** [reply]
Okay, but what if I wand to listen to one port, and connect to other, or both listen and send to this port.

**Reply by svpcom** [reply]
> Okay, but what if I wand to listen to one port, and connect to other, or both listen and send to this port.

Please explain your setup. When listening on UDP port mavlink router will send replies back to all discovered peers.

**Reply by SzateX** [reply]
I have one drone, and one ground station. Thay have wifibroadcast configured and use one Alfa NIC each. I have own software (which is using linux sockets for mavlink) on drone and gs. Drone send video, and mavlink messages. GS send mavlink commands. Its all.

**Reply by SzateX** [reply]
I'm trying to use mavlink router, but mavlink router on drone can't also bind to 14550 port to route it to another port. Setting connect on both drone and gs doesn't help. The messages don't work in any direction.

So how to do bidirectional mavlink communication?

**Reply by svpcom** [reply]
You should got something like:
```
drone <- UART -> mavlink_router <- udp -> WFB_drone <-- radio --> WFB_gs <- udp -> qgroundcontrol
```
mavlink_router has two modes: 
1. Listen on port (udp 14550 by default)
2. Connect to endpoint ( "-e" command-line argument)
See its docs for reference

So if mavlink-router listen port then WFB should connect to it and vice versa

**Reply by SzateX** [reply]
I think you don't read what I'm writting. I'm not using QGroundControl. I'm writting own soft, and don't think now about UART. There is no problem for me. My soft on drone can't recive messages from gs, because on drone I can't bind to port 14550 (it is used by your soft.

**Reply by SzateX** [reply]
Trying to run both services (gs and drone) at the same time with different ports cause closing one of them.

**Reply by svpcom** [reply]
If you write your own soft you should be familiar how mavlink routing works. If you want to use greater than two mavlink peers on drone (include drone autopilot itself) you need to use mavlink router (which will route mavlink packets from WFB, drone and your program). Bind drone, WFB and your program to different UDP ports and add them as endpoints to mavlink router.

**Reply by SzateX** [reply]
Okay. If I have in /etc/wifibroadcast.cfg on drone:
```
[drone_mavlink]
peer = 'listen://0.0.0.0:14550
```

And want to use mavlink_router on drone with this command:

```
mavlink-routerd -e 127.0.0.1:14750  0.0.0.0:14550
``` 

to route packets which are sent from my gs into drone, to my app (Bind on port 14750). I got error: Port 14550 is in use.

**Reply by svpcom** [reply]
No. In this config mavlink-router and WFB both listen 14550.
You need to use `mavlink-routerd -e 127.0.0.1:14750  -e 127.0.0.1:14550 0.0.0.0:14551` or use
```
[drone_mavlink]
peer = 'connect://127.0.0.1:14550
```

**Reply by svpcom** [reply]
Also don't connect several peers to one UDP port - only one will work because mavlink use last ip address to send replies to incoming packets

**Reply by SzateX** [reply]
But If i put connect into config i'm unable to send messages to gs.

**Reply by svpcom** [reply]
Use `listen` in config on drone side, because if you have `connect` on the both ends you can not be able to establish mavlink connection because WFB doesn't emit new packets (only transmit them) and GS and drone's mavlink router will deadlock (because they both are listen for peers)

**Reply by svpcom** [reply]
Run `mavlink-routerd -vv` to debug packets flow

**Reply by SzateX** [reply]
> 
> 
> Use `listen` in config on drone side, because if you have `connect` on the both ends you can not be able to establish mavlink connection because WFB doesn't emit new packets (only transmit them) and GS and drone's mavlink router will deadlock (because they both are listen for peers)

Decide. Listen or Connect.



> 
> 
> No. In this config mavlink-router and WFB both listen 14550.
> You need to use `mavlink-routerd -e 127.0.0.1:14750 -e 127.0.0.1:14550 0.0.0.0:14551` or use
> 
> ```
> [drone_mavlink]
> peer = 'connect://127.0.0.1:14550
> ```

If i run this mavlink-router command it still doesn't work. Sending packets from drone (drone is writting to: 127.0.0.1:14550) to gs (gs is listening 0.0.0.0:14550) works. From gs (gs is writting to 127.0.0.1:14550) to drone (drone is listening 0.0.0.0:14750 and `mavlink-routerd -vv -e 127.0.0.1:14750 -e 127.0.0.1:14550 0.0.0.0:14551` applied). No output on mavlink-router.

**Reply by svpcom** [reply]
GS should not write to 127.0.0.1:14550, but should use source ip and port of received udp packets to send replies. See mavlink router sources for reference.

**Reply by SzateX** [reply]
Now if I do like you say, I got one message from drone to gs. And all communication freezes (I need to restart service to go again)

**Reply by svpcom** [reply]
If connect QGroundControl in parallel to your app - is it working?
